### ISSUE TITLE
Breaking changes in the 1.2.162 headers.

### ISSUE THREAD

**Issue by Kangz** [KhronosGroup/Vulkan-Headers#167]
The 1.2.162 headers contain breaking changes that remove all the enums for the provisional ray-tracing extension. This makes it **extremely** painful to roll the headers in existing projects because it prevents gradual migration to the new API (even just compilation-wise). This gets even worse for projects that have dependencies that use use the enums from the provisional extension (like the VVL).

Can the WG consider using a `_PROVISIONAL_` naming scheme next time and if it wants to remove the enums consider deprecating them over time (like ~a year)?

**Comment by Kangz** [comment]
Same thing for 1.2.157 and the removal of `VK_STRUCTURE_TYPE_DRM_FORMAT_MODIFIER_PROPERTIES_EXT `

**Comment by dgkoch** [comment]
This is exactly why we put the provisional extensions in the vulkan_beta.h header and came up with an opt-in platform define, and told people not to use them in production code bases. Obviously the enums didn't make it into the beta header, but how or why were existing projects using them if they weren't trying to add provisional support?
Yes VVL actually used the provisional headers, but that was a choice we had the changes for those all prepared and staged to push out ASAP (so it's a little frustrating that they still haven't landed). 

I don't think this issue tracker is really the right place for this discussion though.

**Comment by Kangz** [comment]
The enums weren't in the `vulkan_beta.h` header and are causing at least some of the issues I've seen: https://github.com/KhronosGroup/Vulkan-Headers/pull/165/files#diff-e222ae95c2b0d5082b94d6086fb1c24da18ee31384c1a39840df3b9152023ee6L441

I'm glad that you have a VVL change mostly ready to go, that's what's blocking us the most at the moment. That said even if Khronos coordinated to land the changes in all parts of the ecosystem it controls at the same time, some form of staged deprecation would be better. Right now we are going to have to go through the dependency tree of Chromium and carefully roll multiple dependencies at once, and that multiple times as we go down the dependency tree towards Chromium.

Having left the enums and the vulkan_beta.h header, albeit marked as deprecated would have helped provide some overlap and allowed upgrading dependencies one by one instead of what we have to do now.

**Comment by dgkoch** [comment]
I'm unclear why changing the enums would have an effect if the components aren't using them - are there some components that try to do something with every enum listed in the header?  Is it the textual spelling that matters, or the values of them?

> Having left the enums and the vulkan_beta.h header, albeit marked as deprecated would have helped provide some overlap and allowed upgrading dependencies one by one instead of what we have to do now.

From a practical perspective - I don't know how we could have had both the provisional (in the beta header) and the final (in the main header) semantics being driven off of the same XML file, without some more drastic changes to the infrastructure. 
Perhaps if we could have kept the enums in the beta header as well, that might have made the transition easier (but that has it's own set of complexities and issues).

**Comment by Kangz** [comment]
> I'm unclear why changing the enums would have an effect if the components aren't using them - are there some components that try to do something with every enum listed in the header? Is it the textual spelling that matters, or the values of them?

The issue is that many (most?) Vulkan project depend on ecosystem dependencies that take advantage of many enum values (SPIRV-Tools, VVLs, glslang?, spirv-cross). It means that the project has to upgrade all dependencies past the breaking change at the same time, something that is scary given we oftentimes have to upstream build fixes for the upgrades. This gets worse when your project itself depends on projects that depend on the ecosystem.

For example Dawn depends on Tint that depends on spirv-tools. The following happened:

 - As part of a Tint task a PR was merged in SPIRV-Tools after the breaking change
 - Tint rolls both SPIRV-Tools and spirv-headers past the breaking change (which looks painless so far)
 - The roll of Tint into Dawn start failing because we're missing the SPIRV-Tools PR
 - To be able to roll Tint, Dawn needs to roll spirv-headers and SPIRV-Tools but also SPIRV-Cross and other parts of the ecosystem. **this is likely to break**
 - Same thing rolling Dawn into Chromium, and all of the above is also happening in various forms for ANGLE, Swiftshader, Skia, ...

Hopefully this motivates the need to avoid breaking compilation changes and instead do a gradual deprecation over some period of time.

> I'm unclear why changing the enums would have an effect if the components aren't using them - are there some components that try to do something with every enum listed in the header? Is it the textual spelling that matters, or the values of them?

It is the textual spelling, the values don't matter they are unlikely to cause compilation errors when they change.

My suggestion would be that provisional extensions can use the `KHRX` suffix instead of `KHR` similarly to `NV` and `NVX`. It makes it clear in the name that this is Xperimental, and doesn't conflict with any name changes later down the line. It has the drawback that users of the experimental extension that move to the final extension will have to update all names, but at least in the case of Vulkan ray tracing, that's nothing compared to the rest of the updates required.

Process-wise I suggest removing the `KHRX` things only after a decently long deprecation period that's in the order of months so that all parts of the ecosystem have time to transition away from them.

**Comment by nanonyme** [comment]
Is this the reason why our ABI checker is claiming there's an ABI break with enums in 1.2.162? https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/merge_requests/4168#note_474177858

**Comment by nanonyme** [comment]
The real problem here seems to be that some enum *values* are being re-used for different purposes. That might result in issues if two pieces of software disagree what an integer value in an enum means.

**Comment by dgkoch** [comment]
> Is this the reason why our ABI checker is claiming there's an ABI break with enums in 1.2.162? https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/merge_requests/4168#note_474177858

Yes - enums (and the ABI) between the **provisional** ray tracing extensions and the **final** ray tracing extensions intentionally changed (and this was well telegraphed in advance). I believe that all the enums changes your script detected in https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/merge_requests/4168 are as expected. 

In normal (ie non-provisional) cases, the ABI should never break.

**Comment by dgkoch** [comment]
The WG has the following proposal for this going forward. 

1. add a new "protect" xml attribute. 
2. when used with an enum an additional #ifdef symbol will be placed around the enum
3. use this for all enums defined on provisional extensions to protect their definition in vulkan_core.h with 
`VK_ENABLE_BETA_EXTENSIONS`

So for example - with the only current provisional extension `VK_KHR_portability_subset` we would make the following XML change: 
```
-                <enum offset="0" extends="VkStructureType"                  name="VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PORTABILITY_SUBSET_FEATURES_KHR"/>
-                <enum offset="1" extends="VkStructureType"                  name="VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PORTABILITY_SUBSET_PROPERTIES_KHR"/>
+                <enum offset="0" extends="VkStructureType"                  name="VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PORTABILITY_SUBSET_FEATURES_KHR" protect="VK_ENABLE_BETA_EXTENSIONS"/>
+                <enum offset="1" extends="VkStructureType"                  name="VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PORTABILITY_SUBSET_PROPERTIES_KHR" protect="VK_ENABLE_BETA_EXTENSIONS"/>
```
Which results in the following header change: 

```
+#ifdef VK_ENABLE_BETA_EXTENSIONS
     VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PORTABILITY_SUBSET_FEATURES_KHR = 1000163000,
+#endif
+#ifdef VK_ENABLE_BETA_EXTENSIONS
     VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PORTABILITY_SUBSET_PROPERTIES_KHR = 1000163001,
+#endif
```

@Kangz @nanonyme (and anyone else who cares): 
1. Provided we use this for any new provisional extensions going forward, do you think this would adequately protect your use cases if any of the provisional enums are renamed or removed?
2. If we apply this to the already existing portability extension - this will likely have the same effect as removing the tokens - will that cause problems for your dependencies?  If so, could you work around them by temporarily by defining VK_ENABLE_BETA_EXTENSIONS? (but obviously removing it after the dependencies have been updated, or you'll end up with the same problem again!)

**Comment by nanonyme** [comment]
Note for *us* much of the impact was getting false positive complaints (which based on your explanation this was) from ABI checker. This is possible to override once confirmed upstream. If above means provisional things will only show up once more in our checker as violation and not after that there are no more false positives, it sounds reasonable.
